File: Project/Project3/TicketManager.cpp
#include "TicketManager.h"
#include "AdvancedTicket.h"
#include "GeneralTicket.h"

TicketManager::TicketManager(const string& name, const int number)
	:m_name(name), totalNumber(number)
{
	if (totalNumber > 0)
		ticket = new Ticket * [totalNumber];
	else
		ticket = nullptr;
}

TicketManager::~TicketManager()
{
	if (ticket != nullptr) {
		for (size_t i = 0; i < count; i++) {
			delete ticket[i];
		}
		delete[] ticket;
	}
}

void TicketManager::buy(Ticket* t)
{
	if (count < totalNumber) {
		ticket[count++] = t;
	}
}

void TicketManager::show() const
{
	cout << "-----------  콘서트  정보  -----------" << endl;
	cout << "콘서트 이름: " << m_name <<endl;
	cout << "구매개수: " << count << endl;
	cout << "----------  티켓 예약 현황  ----------\n" << endl;

	int price = 0;
	for (size_t i = 0; i < count; i++) {
		ticket[i]->show();
		cout << endl;
		price += ticket[i]->getPrice();
	}
	cout << "총 구매 금액: " << price << endl;
	cout << "--------------------------------------\n" << endl;
}

void TicketManager::showGeneralTIcket(const bool& card)
{
	cout << "----------  일반 예약 현황  ----------\n" << endl;
	cout << "카드 결제 여부:" << boolalpha << card << endl;
	cout << "--------------------------------------\n" << endl;

	int price = 0;

	for (size_t i = 0; i < count; i++) {
		GeneralTicket* GT = dynamic_cast<GeneralTicket*>(ticket[i]);
		if (GT != nullptr) {
			if (GT->getPayByCredit() == card) {
				ticket[i]->show();
				cout << endl;
				price += ticket[i]->getPrice();
			}
		}
	}
	cout << "--------------------------------------" << endl;
	cout << "총 예약 금액: " << price << endl;
	cout << "--------------------------------------\n" << endl;
}

void TicketManager::showAdvancedTicket()
{
	cout << "----------  사전 예약 현황  ----------\n" << endl;

	int price = 0;

	for (size_t i = 0; i < count; i++) {
		AdvancedTicket* AT = dynamic_cast<AdvancedTicket*>(ticket[i]);
		if (AT != nullptr) {
			ticket[i]->show();
			cout << endl;
			price += ticket[i]->getPrice();
		}
	}

	cout << "--------------------------------------" << endl;
	cout << "총 예약 금액: " << price <<endl;
	cout << "--------------------------------------\n" << endl;
}

ostream& operator<<(ostream& out, const TicketManager& manager)
{
	manager.show();
	return out;
}

File: JavaLecture_src/Week05lab/TicketManager.java
package Week05lab;

public class TicketManager {
	Ticket[] tickets;
	int number;
  	int count=0;
	String name;
	
	
	public TicketManager(String name, int number) {
		super();
		this.number = number;
		this.name = name;
		tickets = new Ticket[this.number];
	}
	
	public void showGeneralTicket(boolean payByCredit) {
		for(Ticket ticket:tickets) {
			if(ticket!=null & ticket instanceof GeneralTicket) {
				GeneralTicket t = (GeneralTicket)ticket;
				if(t.isPayByCredit()==payByCredit) {
					System.out.println(t.toString());
				}
			}
		}
	}
	
	public void showAdvancedTicket(int day) {
		for(Ticket ticket:tickets) {
			if(ticket!=null & ticket instanceof AdvancedTicket) {
				AdvancedTicket t = (AdvancedTicket)ticket;
				if(t.getAdvancedDays()>day) {
					System.out.println(t.toString());
				}
			}
		}
		System.out.println("--------------------------");
	}
	
	public void register(Ticket ticket) { //다형성
		if(this.count < this.number)
			this.tickets[this.count++] = ticket;
		else
			System.out.println("티켓 판매 완료");
	}
	
	public double getTotal() {
		double total = 0.0;
		for(Ticket ticket:tickets) {
			if(ticket != null) {
				total+=ticket.getPrice();
			}else
				break;
		}
		return total;
	}

	@Override
	public String toString() {
		// TODO Auto-generated method stub
		String str = "공연명: "+this.name+"\n";
		str+="좌석수: "+this.number+"\n";
		str+="총 판매티켓 수: "+this.count+"\n";
		str+="--------------------\n";
		for(Ticket ticket:tickets) {
			if(ticket!=null)
				str+= ticket.toString() + "\n---------------\n";
			else
				break;
		}
		str+="총 티켓 판매 금액: "+this.getTotal()+"\n";
		
		return str;
	}
	
	


}
